File: src/main/java/com/edgelab/hospital/Patient.java
package com.edgelab.hospital;

import java.util.List;
import java.util.function.BiFunction;

public class Patient {

    private State state;

    private boolean treated = false;

    public Patient(State disease) {
        this.state = disease;
    }

    public State getState() {
        return state;
    }

    void treat(List<Drug> drugs, BiFunction<List<Drug>, State, State> drugEffect) {
        if (this.treated) return;
        var newState = drugEffect.apply(drugs, this.state);
        if (!newState.equals(this.state)) {
            this.state = newState;
            this.treated = true;
        }
    }

}

File: src/main/kotlin/com/edgelab/hospital/Patient.kt
package com.edgelab.hospital

class Patient(var state: State) {

    private var isTreated: Boolean = false

    fun treat(drugs: List<Drug>) {
        Effect.values() // the order is important here
            .onEach { effect -> effect.drugs = drugs }
            .filter { !isTreated }
            .map { effect -> effect.apply(state) }
            .firstOrNull { newState -> newState != state }
            ?.also { newState -> state = newState; isTreated = true }
        isTreated = false
    }

}
